src/actions/channel_categories.ts

```diff
@@ -234,6 +234,85 @@ export function moveChannelToCategory(categoryId: string, channelId: string, new
     };
 }
 
+export function moveChannelsToCategory(categoryId: string, channelIds: string[], newIndex: number, setManualSorting = true) {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        const state = getState();
+        const targetCategory = getCategory(state, categoryId);
+        const currentUserId = getCurrentUserId(state);
+
+        // Add the channel to the new category
+        let categories = {
+            [targetCategory.id]: {
+                ...targetCategory,
+                sorting: (setManualSorting && targetCategory.type !== CategoryTypes.DIRECT_MESSAGES) ? CategorySorting.Manual : targetCategory.sorting,
+                channel_ids: insertMultipleWithoutDuplicates(targetCategory.channel_ids, channelIds, newIndex),
+            },
+        };
+
+        // Needed if we have to revert categories and for checking for favourites
+        let unmodifiedCategories = {[targetCategory.id]: targetCategory};
+        let sourceCategories: Record<string, string> = {};
+
+        // And remove it from the old category
+        channelIds.forEach((channelId) => {
+            const sourceCategory = getCategoryInTeamWithChannel(getState(), targetCategory.team_id, channelId);
+            if (sourceCategory && sourceCategory.id !== targetCategory.id) {
+                unmodifiedCategories = {
+                    ...unmodifiedCategories,
+                    [sourceCategory.id]: sourceCategory,
+                };
+                sourceCategories = {...sourceCategories, [channelId]: sourceCategory.id};
+                categories = {
+                    ...categories,
+                    [sourceCategory.id]: {
+                        ...(categories[sourceCategory.id] || sourceCategory),
+                        channel_ids: removeItem((categories[sourceCategory.id] || sourceCategory).channel_ids, channelId),
+                    },
+                };
+            }
+        });
+
+        const categoriesArray = Object.values(categories).reduce((allCategories: ChannelCategory[], category) => {
+            allCategories.push(category);
+            return allCategories;
+        }, []);
+
+        const result = dispatch({
+            type: ChannelCategoryTypes.RECEIVED_CATEGORIES,
+            data: categoriesArray,
+        });
+
+        try {
+            await Client4.updateChannelCategories(currentUserId, targetCategory.team_id, categoriesArray);
+        } catch (error) {
+            forceLogoutIfNecessary(error, dispatch, getState);
+            dispatch(logError(error));
+
+            const originalCategories = Object.values(unmodifiedCategories).reduce((allCategories: ChannelCategory[], category) => {
+                allCategories.push(category);
+                return allCategories;
+            }, []);
+
+            dispatch({
+                type: ChannelCategoryTypes.RECEIVED_CATEGORIES,
+                data: originalCategories,
+            });
```

Comment:
not sure if we shouldn't trigger a request for the categories from the server, if there was an error there is the possibility that client's info is not right.

Reply:
This is just replacing the categories that we updated above outside of the `catch` back to what they originally were. We can assume the client data that existed before any updates happened is correct since there was no error leading up to that data existing.